File: util/src/main/java/com/github/nsc/de/shake/util/ArrayUtil.java
package com.github.nsc.de.shake.util;

import java.lang.reflect.Array;
import java.util.function.Function;

public class ArrayUtil {

    @SuppressWarnings("unchecked")
    public static <S,T> T[] map(S[] array, T[] target, Function<S, T> fun) {
        if(target.length < array.length) target = (T[]) Array.newInstance(target.getClass().getComponentType(), array.length);
        for(int i = 0 ; i < array.length; i++) target[i] = fun.apply(array[i]);
        return target;
    }

    @SuppressWarnings("unchecked")
    public static <T> T[] map(byte[] array, T[] target, Function<Byte, T> fun) {
        if(target.length < array.length) target = (T[]) Array.newInstance(target.getClass().getComponentType(), array.length);
        for(int i = 0 ; i < array.length; i++) target[i] = fun.apply(array[i]);
        return target;
    }
}

File: util/src/main/kotlin/com/github/nsc/de/shake/util/ArrayUtil.kt
package com.github.nsc.de.shake.util

import java.util.function.Function

object ArrayUtil {

    @JvmStatic
    fun <S, T> map(array: Array<S>, target: Array<T>, f: Function<S, T>): Array<T> {
        var target = target
        if (target.size < array.size) target =
            java.lang.reflect.Array.newInstance(target.javaClass.componentType, array.size) as Array<T>
        for (i in array.indices) target[i] = f.apply(array[i])
        return target
    }

    @JvmStatic
    fun <T> map(array: ByteArray, target: Array<T>, f: Function<Byte?, T>): Array<T> {
        var target = target
        if (target.size < array.size) target =
            java.lang.reflect.Array.newInstance(target.javaClass.componentType, array.size) as Array<T>
        for (i in array.indices) target[i] = f.apply(array[i])
        return target
    }
    
}